### PR TITLE
[WEB-3770] - TIDE Dashboard Delta Summary for Patient Drawer

### PR DESCRIPTION
**Pull request by henry-tp**
This PR is much smaller than the line count suggests! A lot of it is test migrations that are mostly 1:1 but I had to change the assertion syntax and such.

The decision to generate a second AGP versus getting stats numbers directly from the backend summary was an intentional one. What I realized is that the small discrepancies from the differences in the AGP calculation method vs the Backend Summary calculation method show up in a pretty visible way, since we are highlighting single-digit values for the delta (ie. `Increased by 2%`). We don't want an output like `Time CGM Active: Was 82.2%, Current 86.4%, Increased by 2.8%`. The only way to ensure we're comparing apples to apples and guaranteeing an output that is arithmetically consistent is to generate the last period's AGP, then do then comparison between AGPs.

image

### CODE REVIEW

File: __tests__/unit/app/pages/dashboard/PatientDrawer/useAgpCGM/buildGenerateAGPImages.test.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
+/* global jest */
```
Comment (henry-tp) [inline __tests__/unit/app/pages/dashboard/PatientDrawer/useAgpCGM/buildGenerateAGPImages.test.js:1]: This is not a new test file, it's a port of the old one, github is just not recognizing it as such

File: __tests__/unit/app/pages/dashboard/PatientDrawer/Content.test.js [hunk 2]
```diff
@@ -0,0 +1,156 @@
+/* global jest */
+/* global expect */
```
Comment (henry-tp) [inline __tests__/unit/app/pages/dashboard/PatientDrawer/Content.test.js:2]: This is not a new test file, it's a port of the old one, github is just not recognizing it as such

File: app/pages/dashboard/PatientDrawer/useAgpCGM/getOpts.js [hunk 3]
```diff
@@ -53,21 +53,27 @@ const getOpts = (
 
   const dates = getLastNDays(agpPeriodInDays, 'agpCGM');
 
+  const offsetDates = {
+    startDate: dates.startDate.clone().subtract(agpPeriodInDays, 'days'),
+    endDate: dates.endDate.clone().subtract(agpPeriodInDays, 'days'),
+  };
```
Comment (henry-tp) [inline app/pages/dashboard/PatientDrawer/useAgpCGM/getOpts.js:59]: My brain is confused - could it really be this easy? It seems too easy given how much date calculation is going on above. A second scrutinizing eye would be highly appreciated.

File: app/pages/dashboard/PatientDrawer/useAgpCGM/useAgpCGM.js [hunk 4]
```diff
@@ -112,11 +112,12 @@ const useAgpCGM = (
...
 
-  return { 
-    status:      lastCompletedStep, 
-    svgDataURLS: isCorrectPatientInState ? pdf.opts?.svgDataURLS : null,
-    agpCGM:      isCorrectPatientInState ? pdf.data?.agpCGM : null,
+  return {
+    status:       lastCompletedStep,
+    svgDataURLS:  isCorrectPatientInState ? pdf.opts?.svgDataURLS : null,
+    agpCGM:       isCorrectPatientInState ? pdf.data?.agpCGM : null,
+    offsetAgpCGM: isCorrectPatientInState ? pdf.data?.offsetAgpCGM : null,
```
Comment (henry-tp) [inline app/pages/dashboard/PatientDrawer/useAgpCGM/useAgpCGM.js:119]: I would agree if you think `offsetAgpCGM` is not a very good name. I chose the `offset` prefix because it seems to be what the Backend folks call it and what it's called in backend outputs. So it might be meaningful in that context. If I didn't know that I probably would have used `pastAgpCGM` or `pastPeriodAgpCGM`. Open to suggestions or we can stick with the name.

File: __tests__/unit/app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.test.js [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+      expect(screen.queryByTestId('cgm-delta-summary-time-in-very')).not.toBeInTheDocument();
+      expect(screen.queryByTestId('cgm-delta-summary-time-cgm-active')).not.toBeInTheDocument();
+    });
+  });
+
+  describe('When data is present', () => {
+    it('renders the data in the expected format', () => {
+      render(<CGMDeltaSummary agpCGM={agpCGM} offsetAgpCGM={offsetAgpCGM} />);
+
+      const veryLowContainer = screen.queryByTestId('cgm-delta-summary-very-low');
```
Comment (henry-tp) [inline __tests__/unit/app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.test.js]: RTL generally discourages querying by TestId, but I actually think there is value for us, because we don't just want to assert "The number 6 shows up", we want to say "The number 6 shows up _in the right box/row_"

File: app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.js [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+      },
+    },
+  } = offsetAgpCGM;
+
+  const timezone = getTimezoneFromTimePrefs(timePrefs);
+  const dateRange  = getReportDaysText(newestDatum, oldestDatum, bgDaysWorn, timezone);
+  const roundedBgDaysWorn = bankersRound(bgDaysWorn, 0);
+
+  // Current Period Values
+  const timeInVeryHighFraction = _.toNumber(counts.veryHigh) / counts.total * 1;
```
Comment (henry-tp) [inline app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.js]: This is lifted from this line in viz: https://github.com/tidepool-org/viz/blob/64d27d4b17f9d227f73225d805aa120fa67382bd/src/modules/print/utils/AGPUtils.js#L251

I did it for consistency but I'm not entirely sure why there is a `* 1`. Maybe its a shortcut to coerce from string -> number?

File: jest.config.js [hunk 7]
```diff
@@ -2,6 +2,7 @@ module.exports = {
   testEnvironment: 'jsdom',
   roots: ['<rootDir>/__tests__'],
   moduleNameMapper: {
+    '^@app/(.*)$': '<rootDir>/app/$1',
```
Comment (henry-tp) [inline jest.config.js:5]: We can use absolute imports (with an alias) instead of relative imports now

File: app/pages/dashboard/PatientDrawer/useAgpCGM/getOpts.js (hunk 3 above)
Comment (henry-tp) [reply app/pages/dashboard/PatientDrawer/useAgpCGM/getOpts.js:59]: Hmm. This method actually has some unexpected consequences but at the same time those consequences end up being the most logically consistent when I think about it. I'm going to make a loom.

File: app/pages/dashboard/PatientDrawer/useAgpCGM/getOpts.js [hunk 8]
```diff
@@ -51,23 +51,32 @@ const getOpts = (
...
 
+  // Get the date range for the current AGP, ending the date of the latest datum
   const dates = getLastNDays(agpPeriodInDays, 'agpCGM');
 
+  // Get the date range for the offset AGP, ending the moment before the start of current AGP
+  const offsetEndDate = dates.startDate.clone().subtract(1, 'ms');
+  const offsetDates = {
+    startDate: offsetEndDate.clone().subtract(agpPeriodInDays, 'days'),
+    endDate: offsetEndDate,
+  };
```
Comment (henry-tp) [inline app/pages/dashboard/PatientDrawer/useAgpCGM/getOpts.js:61]: A second scrutinizing eye would be highly appreciated to make sure I'm not missing any edge cases.

File: __tests__/unit/app/pages/dashboard/PatientDrawer/CGMStatistics/index.test.js [hunk 9]
```diff
@@ -40,7 +37,7 @@ const agpCGM = {
         },
         'sensorUsage': {
           'sensorUsage': 2532600000,
-          'sensorUsageAGP': 99.95264030310206,
+          'sensorUsageAGP': 99.94264030310206,
```
Comment (henry-tp) [inline __tests__/unit/app/pages/dashboard/PatientDrawer/CGMStatistics/index.test.js]: changed it to test that it renders with 1 decimal (99.9% instead of just 100%)

File: app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.js [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+  // Past Period Values
+  const offsetTimeInVeryHighFraction = _.toNumber(offsetCounts.veryHigh) / offsetCounts.total * 1;
+  const offsetTimeInHighFraction = _.toNumber(offsetCounts.high) / offsetCounts.total * 1;
+  const offsetTimeInTargetFraction = _.toNumber(offsetCounts.target) / offsetCounts.total * 1;
+  const offsetTimeInLowFraction = _.toNumber(offsetCounts.low) / offsetCounts.total * 1;
+  const offsetTimeInVeryLowFraction = _.toNumber(offsetCounts.veryLow) / offsetCounts.total * 1;
+
+  const offsetTimeInVeryHighPercent = bankersRound(offsetTimeInVeryHighFraction * 100, 0);
+  const offsetTimeInHighPercent = bankersRound(offsetTimeInHighFraction * 100, 0);
+  const offsetTimeInTargetPercent = bankersRound(offsetTimeInTargetFraction * 100, 0);
```
Comment (henry-tp-public) [inline app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.js:100]: **Code Duplication Issue**: The calculation logic for time-in-range percentages is heavily duplicated between current and offset periods. This violates DRY principles and makes the code harder to maintain.

**Suggestion**: Extract a reusable function like:
```javascript
const calculateTimeInRangePercentages = (counts) => ({
  veryLow: bankersRound((_.toNumber(counts.veryLow) / counts.total) * 100, 0),
  low: bankersRound((_.toNumber(counts.low) / counts.total) * 100, 0),
  target: bankersRound((_.toNumber(counts.target) / counts.total) * 100, 0),
  high: bankersRound((_.toNumber(counts.high) / counts.total) * 100, 0),
  veryHigh: bankersRound((_.toNumber(counts.veryHigh) / counts.total) * 100, 0),
});
```

File: app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.js [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+  font-weight: bold;
+  color: ${vizColors.purple90};
+`;
+
+const Previous = styled(Box)`
+  font-size: 14px;
+  font-style: italic;
+  color: ${vizColors.gray50};
+`;
+
```
Comment (henry-tp-public) [inline app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.js:46]: **Function Too Complex**: This function is doing too much - data extraction, calculations, and formatting. It's 80+ lines long and hard to follow.

**Suggestion**: Break this into smaller, focused functions:
- `extractCurrentPeriodStats(agpCGM)`
- `extractOffsetPeriodStats(offsetAgpCGM)` 
- `calculateDeltas(currentStats, offsetStats, t)`
- `formatDateDisplayData(offsetAgpCGM, timePrefs)`

This would improve readability, testability, and maintainability.

File: app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.js [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+const CGMDeltaSummary = ({ agpCGM, offsetAgpCGM }) => {
+  const { t } = useTranslation();
+
+  const values = useMemo(() => getRenderedValues(agpCGM, offsetAgpCGM, t), [agpCGM, offsetAgpCGM, t]);
+
+  if (!agpCGM) return null;
+
+  if (!offsetAgpCGM) return <InsufficientData />;
+
```
Comment (henry-tp-public) [inline app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.js:158]: **Magic Number**: The hardcoded `24` hours threshold should be a named constant for better maintainability.

**Suggestion**: Define at the top of the file:
```javascript
const MINIMUM_CGM_HOURS = 24;
```
Then use: `const isDataInsufficient = !hoursOfCGMData || hoursOfCGMData < MINIMUM_CGM_HOURS;`

File: app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.js [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+          bgExtents: { newestDatum, oldestDatum, bgDaysWorn },
+          sensorUsage: { sensorUsageAGP: offsetSensorUsageAGPRaw },
+          timeInRange: { counts: offsetCounts },
+        },
+      },
+    },
+  } = offsetAgpCGM;
+
+  const timezone = getTimezoneFromTimePrefs(timePrefs);
+  const dateRange  = getReportDaysText(newestDatum, oldestDatum, bgDaysWorn, timezone);
```
Comment (henry-tp-public) [inline app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.js:75]: **Redundant Operations**: The `* 1` multiplications are unnecessary and confusing. 

**Suggestion**: Remove them:
```javascript
const timeInVeryHighFraction = _.toNumber(counts.veryHigh) / counts.total;
const timeInHighFraction = _.toNumber(counts.high) / counts.total;
// etc.
```

File: app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.js [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+        </Box>
+      </Flex>
+      <Flex sx={{ justifyContent:'space-between', background: vizColors.blue00, padding: 3, borderRadius: '8px' }}>
+        <Category data-testid="cgm-delta-summary-very-low">
+          <Label>{t('Time in Very Low')}</Label>
+          <Delta>{values?.timeInVeryLowPercentDeltaCopy}</Delta>
+          <Previous>{t('Was {{ value }}%', { value: values?.offsetTimeInVeryLowPercent })}</Previous>
+        </Category>
+        <Category data-testid="cgm-delta-summary-low">
+          <Label>{t('Time in Low')}</Label>
```
Comment (henry-tp-public) [inline app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.js:180]: **Unnecessary Optional Chaining**: Since `values` comes from a memoized function that returns either an object or `{}`, the optional chaining (`?.`) is unnecessary here.

**Suggestion**: Use direct property access:
```javascript
<Delta>{values.timeInVeryLowPercentDeltaCopy}</Delta>
<Previous>{t('Was {{ value }}%', { value: values.offsetTimeInVeryLowPercent })}</Previous>
```

File: app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.js [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+    );
+};
+
+const CGMDeltaSummary = ({ agpCGM, offsetAgpCGM }) => {
+  const { t } = useTranslation();
+
+  const values = useMemo(() => getRenderedValues(agpCGM, offsetAgpCGM, t), [agpCGM, offsetAgpCGM, t]);
+
+  if (!agpCGM) return null;
+
```
Comment (henry-tp-public) [inline app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.js:156]: **Inconsistent Error Handling**: The component checks for insufficient data in multiple places with different logic. Consider creating a dedicated validation function.

**Suggestion**: Create a `validateSufficientData(agpCGM, offsetAgpCGM)` function that returns a clear status object indicating what type of insufficiency exists, making the logic more predictable and testable.

File: __tests__/unit/app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.test.js [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+
+      expect(container).toBeEmptyDOMElement();
+    });
+  });
+
+  describe('When data for previous period is not present', () => {
+    it('renders a message for insufficient data', () => {
+      render(<CGMDeltaSummary agpCGM={agpCGM} offsetAgpCGM={undefined} />);
+
+      expect(screen.getByText('Insufficient data to calculate Time in Ranges')).toBeInTheDocument();
```
Comment (henry-tp-public) [inline __tests__/unit/app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.test.js:150]: **Test Coverage Gap**: The tests don't cover important edge cases like insufficient data scenarios.

**Suggestions for additional test cases**:
- When `hoursOfCGMData < 24`
- When calculation functions receive invalid/malformed data
- When timezone handling fails
- Error boundaries for calculation edge cases

This would improve confidence in the component's robustness.

File: app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.js [hunk 17]
```diff
@@ -0,0 +1,209 @@
...
+  return (
+      <>
+        <Flex mb={2} sx={{ justifyContent: 'space-between', fontSize: 1, fontWeight: 'medium', color: vizColors.gray50 }}>
+          <Box>{t('Tidepool Summary: Changes Since Last Time Period')}</Box>
+        </Flex>
+        <Flex sx={{ justifyContent:'center', background: vizColors.blue00, padding: 3, borderRadius: '8px' }}>
+          <Previous>{t('Insufficient data to calculate Time in Ranges')}</Previous>
+        </Flex>
+      </>
```
Comment (henry-tp-public) [inline app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.js:146]: **Performance Consideration**: Including `t` (translation function) in the useMemo dependency array could cause unnecessary recalculations if the translation function reference changes.

**Suggestion**: Consider whether `t` actually changes between renders, or move translation calls outside the memoized calculation and only memoize the numerical computations.

**Review by henry-tp-public — CHANGES_REQUESTED** [review]
## Code Review Summary

Overall, this is a solid implementation of the CGM Delta Summary feature. The core functionality is well-designed and the test coverage for the happy path is comprehensive. However, there are several areas where the code quality could be improved:

### 🎯 **Key Strengths**
- Good use of memoization for performance
- Clean separation of UI and business logic
- Comprehensive test coverage for main scenarios
- Proper accessibility with test IDs
- Consistent error messaging

### 🔧 **Main Areas for Improvement**
1. **Code Organization**: The `getRenderedValues` function is too complex and should be broken down
2. **DRY Principle**: Significant duplication in percentage calculations
3. **Error Handling**: Inconsistent validation logic could be consolidated
4. **Test Coverage**: Missing edge case scenarios
5. **Code Quality**: Minor issues with magic numbers and redundant operations

### 💡 **Recommendations**
The changes I've suggested would improve maintainability, testability, and robustness without affecting the core functionality. Most are refactoring improvements that would make future development easier.

**Priority**: I'd recommend addressing the function complexity and code duplication issues first, as these will have the biggest impact on long-term maintainability.

Great work on implementing this feature! The business logic appears sound and the approach of generating separate AGPs for comparison is well-reasoned.

File: app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.js (hunk 6 above)
Comment (krystophv) [reply app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.js]: Looks like it's a vestige of a migration from doing math on a percentage value vs. its raw value here: https://github.com/tidepool-org/viz/commit/e5025f5f15326e824ec584a88b645640d8347a5e and that, due to operator precedence, the division will happen before the multiplication anyway, making the `* 1` entirely unnecessary (unless I've missed something). We're probably better off not inheriting that.

File: app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.js [hunk 18]
```diff
@@ -0,0 +1,209 @@
...
+
+  const values = useMemo(() => getRenderedValues(agpCGM, offsetAgpCGM, t), [agpCGM, offsetAgpCGM, t]);
+
+  if (!agpCGM) return null;
+
+  if (!offsetAgpCGM) return <InsufficientData />;
+
+  const { count, sampleInterval } = offsetAgpCGM?.data?.current?.stats?.sensorUsage || {};
+  const hoursOfCGMData = (count * sampleInterval) / MS_IN_HOUR;
+  const isDataInsufficient = !hoursOfCGMData || hoursOfCGMData < 24;
```
Comment (krystophv) [inline app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.js]: Might want to avoid the magic number here (even though it's fairly obvious what it's about)

File: app/pages/dashboard/PatientDrawer/Content.js [hunk 19]
```diff
@@ -1,22 +1,23 @@
 import React from 'react';
+import _ from 'lodash';
```
Comment (krystophv) [inline app/pages/dashboard/PatientDrawer/Content.js]: This seems unused

File: app/pages/dashboard/PatientDrawer/Content.js [hunk 20]
```diff
@@ -1,22 +1,23 @@
...
 import { useSelector } from 'react-redux';
 import { useTranslation } from 'react-i18next';
 import { Flex, Box, Text } from 'theme-ui';
-import colorPalette from '../../../themes/colorPalette';
-import { colors as vizColors } from '@tidepool/viz';
+import { utils as vizUtils, colors as vizColors } from '@tidepool/viz';
```
Comment (krystophv) [inline app/pages/dashboard/PatientDrawer/Content.js]: `vizUtils` seems unused

File: app/pages/dashboard/PatientDrawer/useAgpCGM/getOpts.js (hunk 8 above)
Comment (krystophv) [reply app/pages/dashboard/PatientDrawer/useAgpCGM/getOpts.js:61]: Looks reasonable to me 👍🏼

File: app/pages/dashboard/PatientDrawer/useAgpCGM/useAgpCGM.js (hunk 4 above)
Comment (krystophv) [reply app/pages/dashboard/PatientDrawer/useAgpCGM/useAgpCGM.js:119]: I think adopting the backend nomenclature is reasonable here 👍🏼

File: __tests__/unit/app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.test.js [hunk 21]
```diff
@@ -0,0 +1,203 @@
...
+          total: 3975,
+        },
+        glucoseManagementIndicator: {
+          glucoseManagementIndicator: 6.525028612671605,
+          glucoseManagementIndicatorAGP: 6.525028612671605,
+          total: 3975,
+        },
+        sensorUsage: {
+          sensorUsage: 1192500000,
+          sensorUsageAGP: 98.21119996031549,
```
Comment (krystophv) [inline __tests__/unit/app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.test.js]: value literal technically exceeds double float precision

File: __tests__/unit/app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.test.js [hunk 22]
```diff
@@ -0,0 +1,203 @@
...
+      render(<CGMDeltaSummary agpCGM={agpCGM} offsetAgpCGM={undefined} />);
+
+      expect(screen.getByText('Insufficient data to calculate Time in Ranges')).toBeInTheDocument();
+
+      expect(screen.queryByTestId('cgm-delta-summary-time-in-very-low')).not.toBeInTheDocument();
+      expect(screen.queryByTestId('cgm-delta-summary-time-in-low')).not.toBeInTheDocument();
+      expect(screen.queryByTestId('cgm-delta-summary-time-in-target')).not.toBeInTheDocument();
+      expect(screen.queryByTestId('cgm-delta-summary-time-in-high')).not.toBeInTheDocument();
+      expect(screen.queryByTestId('cgm-delta-summary-time-in-very')).not.toBeInTheDocument();
+      expect(screen.queryByTestId('cgm-delta-summary-time-cgm-active')).not.toBeInTheDocument();
```
Comment (krystophv) [inline __tests__/unit/app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.test.js]: I think these test IDs are incorrect and, as a result, will _always_ not be in the document.

File: __tests__/unit/app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.test.js [hunk 23]
```diff
@@ -0,0 +1,203 @@
...
+  describe('When data is present', () => {
+    it('renders the data in the expected format', () => {
+      render(<CGMDeltaSummary agpCGM={agpCGM} offsetAgpCGM={offsetAgpCGM} />);
+
+      const veryLowContainer = screen.queryByTestId('cgm-delta-summary-very-low');
+      const lowContainer = screen.queryByTestId('cgm-delta-summary-low');
+      const targetContainer = screen.queryByTestId('cgm-delta-summary-target');
+      const highContainer = screen.queryByTestId('cgm-delta-summary-high');
+      const veryHighContainer = screen.queryByTestId('cgm-delta-summary-very-high');
+      const cgmActiveContainer = screen.queryByTestId('cgm-delta-summary-cgm-active');
```
Comment (krystophv) [inline __tests__/unit/app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.test.js]: Generally think we should prefer the `getBy*` APIs for elements that _should_ exist.

File: __tests__/unit/app/pages/dashboard/PatientDrawer/CGMStatistics/index.test.js (hunk 9 above)
Comment (krystophv) [reply __tests__/unit/app/pages/dashboard/PatientDrawer/CGMStatistics/index.test.js]: value literal technically exceeds double float precision

File: __tests__/unit/app/pages/dashboard/PatientDrawer/Content.test.js [hunk 24]
```diff
@@ -0,0 +1,156 @@
...
+/* global describe */
+/* global sinon */
+/* global afterEach */
+/* global describe */
```
Comment (krystophv) [inline __tests__/unit/app/pages/dashboard/PatientDrawer/Content.test.js]: double defined global

File: __tests__/unit/app/pages/dashboard/PatientDrawer/Content.test.js [hunk 25]
```diff
@@ -0,0 +1,156 @@
...
+      render(<Provider store={store}> <Content {...props} /> </Provider>);
+
+      expect(screen.getByText('Naoya Inoue does not have any data yet.')).toBeInTheDocument();
+      expect(screen.queryByText('Time in Ranges')).not.toBeInTheDocument();
+      expect(screen.queryByText('Ambulatory Glucose Profile (AGP)')).not.toBeInTheDocument();
+      expect(screen.queryByText('Daily Glucose Profiles')).not.toBeInTheDocument();
+    });
+  });
+
+  describe('shows no data fields and an appropriate message to the user', () => {
```
Comment (krystophv) [inline __tests__/unit/app/pages/dashboard/PatientDrawer/Content.test.js]: Might need to update test description regarding insufficient data

File: __tests__/unit/app/pages/dashboard/PatientDrawer/Content.test.js [hunk 26]
```diff
@@ -0,0 +1,156 @@
...
+
+        expect(percentInRangesImage).toHaveAttribute('src', 'percentInRanges.img.jpg');
+        expect(agpImage).toHaveAttribute('src', 'ambulatoryGlucoseProfile.img.jpg');
+        expect(dailyFirstImage).toHaveAttribute('src', 'daily.top.img.jpg');
+        expect(dailySecondImage).toHaveAttribute('src', 'daily.bot.img.jpg');
+      });
+    });
+
+    describe('When not enough data to render AGP Graph', () => {
+      it('shows the AGP Report with all images', () => {
```
Comment (krystophv) [inline __tests__/unit/app/pages/dashboard/PatientDrawer/Content.test.js]: probably should update test description regarding not showing AGP chart

File: __tests__/unit/app/pages/dashboard/PatientDrawer/useAgpCGM/buildGenerateAGPImages.test.js [hunk 27]
```diff
@@ -0,0 +1,78 @@
...
+/* global Promise */
+/* global expect */
+/* global describe */
+/* global afterEach */
+/* global describe */
```
Comment (krystophv) [inline __tests__/unit/app/pages/dashboard/PatientDrawer/useAgpCGM/buildGenerateAGPImages.test.js]: second `describe` global

File: __tests__/unit/app/pages/dashboard/PatientDrawer/useAgpCGM/buildGenerateAGPImages.test.js [hunk 28]
```diff
@@ -0,0 +1,78 @@
...
+/* global it */
+/* global beforeEach */
+/* global before */
+/* global after */
+
+import React from 'react';
+import { waitFor } from '@testing-library/react';
+import _ from 'lodash';
```
Comment (krystophv) [inline __tests__/unit/app/pages/dashboard/PatientDrawer/useAgpCGM/buildGenerateAGPImages.test.js]: unused import?

File: __tests__/unit/app/pages/dashboard/PatientDrawer/useAgpCGM/buildGenerateAGPImages.test.js [hunk 29]
```diff
@@ -0,0 +1,78 @@
...
+      generateAGPFigureDefinitions: jest.fn(),
+    },
+  },
+}));
+
+describe('buildGenerateAGPImages', () => {
+  const dispatch = jest.fn();
+
+  let toImage;
+  let generateAGPFigureDefinitions;
```
Comment (krystophv) [inline __tests__/unit/app/pages/dashboard/PatientDrawer/useAgpCGM/buildGenerateAGPImages.test.js]: variables are never used

File: __tests__/unit/app/pages/dashboard/PatientDrawer/useAgpCGM/useAgpCGM.test.js [hunk 30]
```diff
@@ -1,15 +1,14 @@
-/* global chai */
+/* global jest */
+/* global expect */
 /* global describe */
-/* global sinon */
 /* global afterEach */
-/* global context */
+/* global describe */
```
Comment (krystophv) [inline __tests__/unit/app/pages/dashboard/PatientDrawer/useAgpCGM/useAgpCGM.test.js]: another dupe describe :)

**Review by krystophv — CHANGES_REQUESTED** [review]
mostly a handful of really small recommended tweaks and one set of test IDs that need to be updated. Overall looking pretty good 🤘🏼

File: app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.js (hunk 6 above)
Comment (henry-tp) [reply app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.js]: Sounds good thank you 9095cf1cd8ed93351c55cee0ea12713bc39a3e39

File: __tests__/unit/app/pages/dashboard/PatientDrawer/CGMStatistics/index.test.js (hunk 9 above)
Comment (henry-tp) [reply __tests__/unit/app/pages/dashboard/PatientDrawer/CGMStatistics/index.test.js]: cut it down a bit e1f1e959ca4e74d641d0c57801741533a68ebbaf

File: app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.js (hunk 18 above)
Comment (henry-tp) [reply app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.js]: Makes sense e1f1e959ca4e74d641d0c57801741533a68ebbaf

File: app/pages/dashboard/PatientDrawer/Content.js (hunk 19 above)
Comment (henry-tp) [reply app/pages/dashboard/PatientDrawer/Content.js]: You are correct e1f1e959ca4e74d641d0c57801741533a68ebbaf

File: app/pages/dashboard/PatientDrawer/Content.js (hunk 20 above)
Comment (henry-tp) [reply app/pages/dashboard/PatientDrawer/Content.js]: You are correct e1f1e959ca4e74d641d0c57801741533a68ebbaf

File: __tests__/unit/app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.test.js (hunk 21 above)
Comment (henry-tp) [reply __tests__/unit/app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.test.js]: I cut it down: 34b535f10

File: __tests__/unit/app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.test.js (hunk 22 above)
Comment (henry-tp) [reply __tests__/unit/app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.test.js]: Haha, this one is pretty egregious. I named it the long one before and thought .. no .. this is way too long. Thank you https://github.com/tidepool-org/blip/commit/e1f1e959ca4e74d641d0c57801741533a68ebbaf

File: __tests__/unit/app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.test.js (hunk 23 above)
Comment (henry-tp) [reply __tests__/unit/app/pages/dashboard/PatientDrawer/CGMDeltaSummary/index.test.js]: Yes, definitely e1f1e959ca4e74d641d0c57801741533a68ebbaf

File: __tests__/unit/app/pages/dashboard/PatientDrawer/Content.test.js (hunk 24 above)
Comment (henry-tp) [reply __tests__/unit/app/pages/dashboard/PatientDrawer/Content.test.js]: Fixed e1f1e959ca4e74d641d0c57801741533a68ebbaf

File: __tests__/unit/app/pages/dashboard/PatientDrawer/Content.test.js (hunk 25 above)
Comment (henry-tp) [reply __tests__/unit/app/pages/dashboard/PatientDrawer/Content.test.js]: Fixed e1f1e959ca4e74d641d0c57801741533a68ebbaf

File: __tests__/unit/app/pages/dashboard/PatientDrawer/Content.test.js (hunk 26 above)
Comment (henry-tp) [reply __tests__/unit/app/pages/dashboard/PatientDrawer/Content.test.js]: Fixed e1f1e959ca4e74d641d0c57801741533a68ebbaf

File: __tests__/unit/app/pages/dashboard/PatientDrawer/useAgpCGM/buildGenerateAGPImages.test.js (hunk 27 above)
Comment (henry-tp) [reply __tests__/unit/app/pages/dashboard/PatientDrawer/useAgpCGM/buildGenerateAGPImages.test.js]: Fixed e1f1e959ca4e74d641d0c57801741533a68ebbaf

File: __tests__/unit/app/pages/dashboard/PatientDrawer/useAgpCGM/buildGenerateAGPImages.test.js (hunk 28 above)
Comment (henry-tp) [reply __tests__/unit/app/pages/dashboard/PatientDrawer/useAgpCGM/buildGenerateAGPImages.test.js]: Fixed e1f1e959ca4e74d641d0c57801741533a68ebbaf

File: __tests__/unit/app/pages/dashboard/PatientDrawer/useAgpCGM/buildGenerateAGPImages.test.js (hunk 29 above)
Comment (henry-tp) [reply __tests__/unit/app/pages/dashboard/PatientDrawer/useAgpCGM/buildGenerateAGPImages.test.js]: Fixed e1f1e959ca4e74d641d0c57801741533a68ebbaf

File: __tests__/unit/app/pages/dashboard/PatientDrawer/useAgpCGM/useAgpCGM.test.js (hunk 30 above)
Comment (henry-tp) [reply __tests__/unit/app/pages/dashboard/PatientDrawer/useAgpCGM/useAgpCGM.test.js]: :) e1f1e959ca4e74d641d0c57801741533a68ebbaf

**Comment by henry-tp** [comment]
Thanks for the review @krystophv , back to you 😄
